tools/httpbintest/src/lib.rs

```diff
@@ -0,0 +1,1304 @@
+
+            let s =
+                match h3_conn.send_request(conn, &req.hdrs, req.body.is_none()) {
+                    Ok(stream_id) => stream_id,
+
+                    Err(e) => {
+                        error!("failed to send request {:?}", e);
+                        return Err(e);
+                    },
+                };
+
+            self.issued_reqs.insert(s, self.current_idx);
+
+            if let Some(body) = &req.body {
+                info!("sending body {:?}", body);
+
+                if let Err(e) = h3_conn.send_body(conn, s, body, true) {
+                    error!("failed to send request body {:?}", e);
+                    return Err(e);
+                }
+            }
+
+            self.current_idx += 1;
+        }
+
+        Ok(())
+    }
+
+    /// Append response headers for an issued request.
+    pub fn add_response_headers(&mut self, stream_id: u64, headers: &[Header]) {
+        let i = self.issued_reqs.get(&stream_id).unwrap();
+        self.reqs[*i].resp_hdrs.extend_from_slice(headers);
+    }
+
+    /// Append data to the response body for an issued request.
+    pub fn add_response_body(
+        &mut self, stream_id: u64, data: &[u8], data_len: usize,
+    ) {
+        let i = self.issued_reqs.get(&stream_id).unwrap();
+        self.reqs[*i].resp_body.extend_from_slice(&data[..data_len]);
+    }
+
+    /// Execute the test assertion(s).
+    pub fn assert(&mut self) -> bool {
+        (self.assert)(&self.reqs)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use std::sync::Once;
+
+    static INIT: Once = Once::new();
+
+    fn run_test<T>(test: T) -> ()
+    where
+        T: FnOnce() -> () + std::panic::UnwindSafe,
+    {
+        INIT.call_once(|| {
+            env_logger::builder()
+                .default_format_timestamp_nanos(true)
+                .init()
+        });
+
+        let result = std::panic::catch_unwind(|| test());
+
+        assert!(result.is_ok())
+    }
+
+    fn endpoint() -> url::Url {
+        let endpoint = match std::env::var_os("HTTPBIN_ENDPOINT") {
+            Some(val) => val.into_string().unwrap(),
+
+            None => String::from("https://cloudflare-quic.com/b"),
+        };
+
+        url::Url::parse(&endpoint).unwrap()
+    }
+
+    fn jsonify(data: &[u8]) -> HttpBinResponseBody {
+        serde_json::from_slice(&data).unwrap()
+    }
+
+    fn do_test(
+        reqs: Vec<HttpBinReq>, assert: HttpBinTestAssert, concurrent: bool,
+    ) {
+        let mut test = HttpBinTest::new(endpoint(), reqs, assert);
+        test.set_concurrency(concurrent);
+        assert!(crate::runner::run(&mut test));
+    }
+
+    // Build a single request and expected response with status code
+    fn request_check_status(testpoint: &str, status: usize) -> Vec<HttpBinReq> {
+        let expect_hdrs = Some(vec![Header::new(":status", &status.to_string())]);
+
+        let mut url = endpoint();
+        let path = format!("{}/{}", url.path(), &testpoint);
```

Comment:
This seems a pretty common operation, maybe `endpoint()` should just take the path as well and do this? Or have some other function.

Reply:
Yeah probably, I'll see what I can do in a refactor.